### ISSUE TITLE
We should have a few default model size parameters for different sizes that can be used.

### ISSUE THREAD

**Issue by mmcdermott** [mmcdermott/MEDS_EIC_AR#56]
Advice on these sourced from claude is as follows:

# Model Sizes
Micro: 10M params (4 layers, 256 hidden, 4 heads, 1024 intermediate)
Small: 50M params (8 layers, 512 hidden, 8 heads, 2048 intermediate)
Medium: 200M params (12 layers, 768 hidden, 12 heads, 3072 intermediate)
Large: 800M params (24 layers, 1024 hidden, 16 heads, 4096 intermediate)

# Batch Size and Learning Rate Strategy for Medical LLM Grid

## Core Philosophy: Scaling Laws Over Individual Tuning

**Goal**: Use established scaling relationships to set hyperparameters systematically rather than tuning each model individually. This ensures fair comparison across your grid.

### Model-Dependent Batch Sizes
Follow the principle: **Batch size should scale with model capacity** to maintain stable gradient estimates.

```
Micro (10M):   Batch size = 64   (effective batch = 64k tokens)
Small (50M):   Batch size = 128  (effective batch = 128k tokens) 
Medium (200M): Batch size = 256  (effective batch = 256k tokens)
Large (800M):  Batch size = 512  (effective batch = 512k tokens)
```

**Rationale**: 
- Larger models need larger batches for stable gradients
- Maintains roughly constant tokens-per-batch scaling
- Avoids memory issues while ensuring statistical stability

### Context-Dependent Adjustments
If training context size is different from 1024 tokens, adjust proportionally:
- **512 context**: Double the batch size
- **2048 context**: Halve the batch size
- **Goal**: Keep total tokens per batch roughly constant

### Scaling Law Approach
Use the established relationship: **Learning rate should scale inversely with model size**

```
Micro (10M):   LR = 3e-4
Small (50M):   LR = 2e-4  
Medium (200M): LR = 1e-4
Large (800M):  LR = 5e-5
```

**Formula**: `LR = base_lr / sqrt(model_size_ratio)`
- Base LR = 3e-4 (for 10M parameter baseline)
- Prevents larger models from unstable training
- Smaller models can handle higher learning rates for faster convergence

### Learning Rate Schedule
**Use identical schedules across all models** for fair comparison:

1. **Warmup**: Linear warmup for first 2000 steps
2. **Decay**: Cosine decay to 10% of peak LR
3. **Total Steps**: Scale with model size
   - Micro: 20k steps
   - Small: 30k steps  
   - Medium: 50k steps
   - Large: 80k steps

### Quick Validation Protocol
Before running your full grid, validate these hyperparameters on 2-3 representative configurations:

1. **Stability Check**: Train for 5k steps, ensure loss decreases smoothly
2. **Learning Rate Sweep**: Try ±50% of recommended LR on one model size
3. **Batch Size Test**: Verify memory usage and training speed

### Warning Signs to Watch For
- **Learning rate too high**: Loss oscillates or explodes after warmup
- **Learning rate too low**: Loss plateaus early, poor final performance
- **Batch size too small**: Noisy training, poor convergence
- **Batch size too large**: Memory issues, or slow gradient updates

## Alternative: Minimal Hyperparameter Search

If you want to be more conservative, run a **very limited grid search** on your smallest model:

```
Learning Rates: [1e-4, 2e-4, 4e-4]
Batch Sizes: [32, 64, 128]
```

Then apply the **relative scaling** to larger models:
- If 2e-4 works best for Micro, use 1e-4 for Medium, 5e-5 for Large
- If batch size 64 works best for Micro, use 256 for Medium, 512 for Large

### Consistent Across All Models
- **Gradient Clipping**: Max norm = 1.0 (prevents exploding gradients)
- **Weight Decay**: 0.1 (standard for transformers)
- **Dropout**: 0.1 (light regularization)

### Why Keep These Constant
- Eliminates confounding variables
- These values are robust across model sizes
- Focus experimental variation on architecture and data

### Mixed Precision Training
**Always use FP16 or BF16** across all experiments:
- Reduces memory usage by ~50%
- Enables larger batch sizes
- Minimal impact on final performance
- Essential for larger models

### Gradient Accumulation
If memory-constrained, use gradient accumulation to achieve target batch sizes:
```python
effective_batch_size = micro_batch_size * accumulation_steps
# Adjust accumulation_steps to hit target effective batch size
```

### Dynamic Batch Sizing
For the largest models, you might need to reduce batch size due to memory constraints:
```python
if model_size > 500M and OOM_error:
    batch_size = batch_size // 2
    accumulation_steps = accumulation_steps * 2
```

### Key Metrics to Track
1. **Training Loss Smoothness**: Should decrease monotonically after warmup
2. **Gradient Norm**: Should be stable (not exploding/vanishing)  
3. **Learning Rate Utilization**: Loss should respond to LR changes
4. **Memory Usage**: Should be <90% of available GPU memory

### Adjustment Protocol
**Only adjust hyperparameters if**:
- Training becomes unstable (loss spikes, NaN values)
- Memory usage exceeds capacity
- Convergence is clearly suboptimal compared to similar configurations

### Documentation
Track all hyperparameter changes and their rationale. This ensures reproducibility and helps identify patterns across your grid.

## Expected Outcomes

With this strategy, you should see:
- **Stable training** across all model sizes
- **Fair comparisons** between configurations
- **Predictable scaling** behavior
- **Efficient compute usage**

The key advantage is that you're using established scaling laws rather than expensive hyperparameter searches, allowing you to focus compute on the scientific questions (tokenization, context size, etc.) rather than optimization details.

# Adam Optimizer Settings
One additional consideration: Adam optimizer settings. I'd recommend keeping these constant across all experiments:

β₁ = 0.9, β₂ = 0.95 (slightly higher β₂ than default, works well for language models)
ε = 1e-8 (standard)
